actors/builtin/miner/miner_test.go

```diff
@@ -86,49 +91,144 @@ func TestConstruction(t *testing.T) {
 	})
 }
 
+// Tests for fetching and manipulating miner addresses.
 func TestControlAddresses(t *testing.T) {
-	actor := actorHarness{miner.Actor{}, t}
 	owner := tutil.NewIDAddr(t, 100)
 	worker := tutil.NewIDAddr(t, 101)
 	workerKey := tutil.NewBLSAddr(t, 0)
 	receiver := tutil.NewIDAddr(t, 1000)
+	actor := newHarness(t, owner, worker, workerKey)
 	builder := mock.NewBuilder(context.Background(), receiver).
 		WithActorType(owner, builtin.AccountActorCodeID).
 		WithActorType(worker, builtin.AccountActorCodeID).
-		WithHasher(blake2b.Sum256).
+		WithHasher(fixedHasher(0)).
 		WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
 
 	t.Run("get addresses", func(t *testing.T) {
 		rt := builder.Build(t)
-
-		actor.constructAndVerify(rt, owner, worker, workerKey)
+		actor.constructAndVerify(rt, miner.WPoStProvingPeriod)
 
 		o, w := actor.controlAddresses(rt)
 		assert.Equal(t, owner, o)
 		assert.Equal(t, worker, w)
 	})
+
+	// TODO: test changing worker (with delay), changing peer id
+}
+
+// Test for sector precommitment and proving.
+func TestCommitments(t *testing.T) {
+	owner := tutil.NewIDAddr(t, 100)
+	worker := tutil.NewIDAddr(t, 101)
+	workerKey := tutil.NewBLSAddr(t, 0)
+	receiver := tutil.NewIDAddr(t, 1000)
+	actor := newHarness(t, owner, worker, workerKey)
+	periodBoundary := abi.ChainEpoch(100)
+	builder := mock.NewBuilder(context.Background(), receiver).
+		WithActorType(owner, builtin.AccountActorCodeID).
+		WithActorType(worker, builtin.AccountActorCodeID).
+		WithHasher(fixedHasher(uint64(periodBoundary))).
+		WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
+
+	t.Run("invalid pre-commit rejected", func(t *testing.T) {
+		rt := builder.Build(t)
+		precommitEpoch := periodBoundary + 1
+		rt.SetEpoch(precommitEpoch)
+		actor.constructAndVerify(rt, periodBoundary+miner.WPoStProvingPeriod)
+		st := getState(rt)
+		deadline, _ := st.DeadlineInfo(precommitEpoch)
+
+		challengeEpoch := precommitEpoch - 20
+
+		// Good commitment.
+		actor.preCommitSector(rt, makePreCommit(100, challengeEpoch, deadline.PeriodEnd()), big.Zero())
+
+		// Duplicate sector ID
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			actor.preCommitSector(rt, makePreCommit(100, challengeEpoch, deadline.PeriodEnd()), big.Zero())
+		})
+
+		// Expires at current epoch
+		rt.SetEpoch(deadline.PeriodEnd())
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			actor.preCommitSector(rt, makePreCommit(111, challengeEpoch, deadline.PeriodEnd()), big.Zero())
+		})
+		rt.SetEpoch(precommitEpoch)
+
+		// Expires before current epoch
+		rt.SetEpoch(deadline.PeriodEnd() + 1)
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			actor.preCommitSector(rt, makePreCommit(112, challengeEpoch, deadline.PeriodEnd()), big.Zero())
+		})
+		rt.SetEpoch(precommitEpoch)
```

Comment:
If this goes with the below `ExpectAbort` (I think it does) it can be moved down next to it.